ddl/job_table.go

```diff
@@ -31,6 +80,145 @@ const (
+			return d.checkJobIsRunnable(sess, sql)
+		}
+		return true, nil
+	})
+}
+
+func (d *ddl) checkJobIsRunnable(sess *session, sql string) (bool, error) {
+	rows, err := sess.execute(context.Background(), sql, "check_runnable")
+	return len(rows) == 0, err
+}
+
+func (d *ddl) getReorgJob(sess *session) (*model.Job, error) {
+	return d.getJob(sess, reorg, func(job *model.Job) (bool, error) {
+		sql := fmt.Sprintf("select job_id from mysql.tidb_ddl_job where schema_id = %d and is_drop_schema and job_id < %d limit 1", job.SchemaID, job.ID)
+		return d.checkJobIsRunnable(sess, sql)
+	})
+}
+
+func (d *ddl) startDispatchLoop() {
+	se, err := d.sessPool.get()
+	if err != nil {
+		logutil.BgLogger().Fatal("dispatch loop get session failed, it should not happen, please try restart TiDB", zap.Error(err))
+	}
+	defer d.sessPool.put(se)
+	sess := newSession(se)
+	var notifyDDLJobByEtcdChGeneral clientv3.WatchChan
+	var notifyDDLJobByEtcdChReorg clientv3.WatchChan
+	if d.etcdCli != nil {
+		notifyDDLJobByEtcdChGeneral = d.etcdCli.Watch(context.Background(), addingDDLJobGeneral)
+		notifyDDLJobByEtcdChReorg = d.etcdCli.Watch(context.Background(), addingDDLJobReorg)
+	}
+	ticker := time.NewTicker(1 * time.Second)
+	defer ticker.Stop()
+	for {
+		if isChanClosed(d.ctx.Done()) {
+			return
+		}
+		if !d.isOwner() {
+			d.once.Store(true)
+			time.Sleep(time.Second)
+			continue
+		}
+		select {
+		case <-d.ddlJobCh:
+		case <-ticker.C:
+		case _, ok := <-notifyDDLJobByEtcdChGeneral:
+			if !ok {
+				logutil.BgLogger().Warn("[ddl] start general worker watch channel closed", zap.String("watch key", addingDDLJobGeneral))
+				notifyDDLJobByEtcdChGeneral = d.etcdCli.Watch(context.Background(), addingDDLJobGeneral)
+				time.Sleep(time.Second)
+				continue
+			}
+		case _, ok := <-notifyDDLJobByEtcdChReorg:
+			if !ok {
+				logutil.BgLogger().Warn("[ddl] start reorg worker watch channel closed", zap.String("watch key", addingDDLJobReorg))
+				notifyDDLJobByEtcdChReorg = d.etcdCli.Watch(context.Background(), addingDDLJobReorg)
+				time.Sleep(time.Second)
+				continue
+			}
+		}
+		d.runDDLJob(sess, d.generalDDLWorkerPool, d.getGeneralJob)
+		d.runDDLJob(sess, d.reorgWorkerPool, d.getReorgJob)
+	}
+}
+
+func (d *ddl) runDDLJob(sess *session, pool *workerPool, getJob func(*session) (*model.Job, error)) {
+	wk, err := pool.get()
+	if err != nil || wk == nil {
+		logutil.BgLogger().Debug(fmt.Sprintf("[ddl] no %v worker available now", pool.tp()), zap.Error(err))
+		return
+	}
+
+	job, err := getJob(sess)
+	if job == nil || err != nil {
+		if err != nil {
+			logutil.BgLogger().Warn("[ddl] get job met error", zap.Error(err))
+		}
+		pool.put(wk)
+		return
+	}
+
+	d.doDDLJob(wk, pool, job)
+}
+
+func (d *ddl) doDDLJob(wk *worker, pool *workerPool, job *model.Job) {
+	injectFailPointForGetJob(job)
+	d.insertRunningDDLJobMap(job.ID)
+	d.wg.Run(func() {
+		defer func() {
+			pool.put(wk)
+			d.deleteRunningDDLJobMap(job.ID)
+			asyncNotify(d.ddlJobCh)
```

Comment:
Why do we need to notify the handler loop **after** the execution of the DDL?

Reply:
let the manager know there is a job that can be run